Tribler/Core/Modules/restapi/events_endpoint.py

```diff
@@ -96,20 +97,25 @@ def __init__(self, session):
         self.session.add_observer(self.on_circuit_removed, NTFY_TUNNEL, [NTFY_REMOVE])
         self.session.add_observer(self.on_search_response, SIGNAL_GIGACHANNEL_COMMUNITY, [SIGNAL_ON_SEARCH_RESULTS])
 
-    def write_data(self, message):
+    def write_data(self, message=None):
         """
         Write data over the event socket if it's open.
         """
-        try:
-            message_str = json.twisted_dumps(message)
-        except UnicodeDecodeError:
-            # The message contains invalid characters; fix them
-            message_str = json.twisted_dumps(fix_unicode_dict(message))
+        if message:
+            self.unsent_messages.append(message)
 
         if len(self.events_requests) == 0:
             return
-        else:
+
+        for m in self.unsent_messages:
```

Comment:
The first design of the events endpoint actually had similar behaviour where messages are queued up if there is no available socket. However, we removed this functionality for the reason that event messages are not important enough to be queued up, and a missing event should not affect the user. In addition, I noticed that the queue size of this implementation is not limited and there is a possibility for memory leaks. @ichorid what is the motivation behind this commit (besides making sure that messages are queued up and flushed to the GUI if there is no socket available)?

Reply:
When TriblerCore starts, it fires the upgrader, which sends 3 messages to the GUI:
1. Upgrade procedure started
2. Upgrader tick (possibly repeated many times)
3. Upgrader finished
This messages should always be sent in order, for upgrade notification/skip procedure to work properly. However, GUI<->Core messaging at startup gives no guarantees on this. For example, message (1) is never seen by the GUI due to some kind of race condition.

To prevent a potential memory leak we can limit the queue size and make it work FIFO.

Reply:
> This messages should always be sent in order, for upgrade notification/skip procedure to work properly.

If possible, I would try to remove any assumptions/requirements on the order of messages being sent/received. Is this possible in this situation?

Reply:
Yes, that is possible, at least for the particular case of the current upgrader logic.
However, I've already redone the thing to use, literally, FIFO queue with limited size.
Do you insist on removing the queue solution and going instead for non-order-dependent-upgrader-notifications solution?

Reply:
In this case, I would suggest to revert this change and not rely on any order if not too much effort 👍